peps/pep-0740.rst

```diff
@@ -0,0 +1,288 @@
+keyservers and key distribution) or (4) (due to PGP signatures typically being
+constructed over just an input file, without any associated signed metadata).
+Similarly, PyPI's historical implementation of PGP did not satisfy consideration
+(5), due to a lack of consistency checks between different release files
+(and an inability to perform those checks due to no access to the signer's
+public key).
+
+Wheel signatures
+^^^^^^^^^^^^^^^^
+
+:pep:`427` (and its :ref:`living PyPA counterpart <packaging:binary-distribution-format>`)
+specify the :term:`wheel format <packaging:Wheel>`.
+
+This format includes accommodations for digital signatures embedded directly
+into the wheel, in either JWS or S/MIME format. These signatures are specified
+over a :pep:`376` RECORD, which is modified to include a cryptographic digest
+for each recorded file in the wheel.
+
+While wheel signatures are fully specified, they do not appear to be broadly
+used; the official `wheel tooling <https://github.com/pypa/wheel>`_ deprecated
+signature generation and verification support
+`in 0.32.0 <https://wheel.readthedocs.io/en/stable/news.html>`_, which was
+released in 2018.
+
+Additionally, wheel signatures do not satisfy any of
+the above considerations (due to the "attached" nature of the signatures,
+non-verifiability on the index itself, and support for wheels only).
+
+Specification
+=============
+
+Upload endpoint changes
+-----------------------
+
+The current upload API is not standardized. However, we propose the following
+changes to it:
+
+* In addition to the current top-level ``content`` and ``gpg_signature`` fields,
+  the index **SHALL** accept ``attestations`` as an additional multipart form
+  field.
+* The new ``attestations`` field **SHALL** be a JSON object.
+* The JSON object **SHALL** have one or more keys, each identifying an
+  attestation format known to the index. If any key does not identify an
+  attestation format known to the index, the index **MUST** reject the upload.
+* The value associated with each well-known key **SHALL** be a JSON object.
+* Each attestation value **MUST** be verifiable by the index. If the index fails
+  to verify any attestation in ``attestations``, it **MUST** reject the upload.
+
+In addition to the above, the index **SHALL** enforce a consistency
+policy for release attestations via the following:
+
+* If the first file under a new release is supplied with ``attestations``,
+  then all subsequently uploaded files under the same release **MUST** also
+  have ``attestations``. Conversely, if the first file under a new release
+  does not have any ``attestations``, then all subsequent uploads under the
+  same release **MUST NOT** have ``attestations``.
+* All files under the same release **MUST** have the same set of well-known
+  attestation format keys.
+
+The index **MUST** reject any file upload that does not satisfy these
+consistency properties.
+
+Index changes
```

Comment:
Should this PEP make any recommendation about how the list of "known keys" for attestations be maintained? Does it take a PEP to create a new key?

Reply:
My intuition here is that these would probably be best suited for the PyPA Specifications (https://www.pypa.io/en/latest/specifications/) rather than PEPs, since they'll be (1) living standards, and (2) potentially very index-specific.

Does that seem reasonable? If so, I can add some language to that effect.

Reply:
This matches what I would recommend as well, thanks!

Reply:
Added a "recommendations" section with this, PTAL!